File: backend/src/main/java/com/storecontrol/backend/models/Voluntary.java
package com.storecontrol.backend.models;

import com.storecontrol.backend.controllers.request.voluntary.RequestCreateVoluntary;
import com.storecontrol.backend.controllers.request.voluntary.RequestUpdateVoluntary;
import jakarta.persistence.*;
import jakarta.validation.Valid;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "volunteers")
@Getter
@NoArgsConstructor
public class Voluntary {
  @Id @GeneratedValue(generator = "UUID")
  private UUID uuid;
  @Embedded
  private User user;
  private String fullname;
  @ManyToOne @JoinColumn(name = "stand_uuid")
  private Stand stand;
  @OneToMany(mappedBy = "voluntary")
  private List<Sale> sales;
  @OneToMany(mappedBy = "voluntary")
  private List<Recharge> recharges;
  @OneToMany(mappedBy = "voluntary")
  private List<Donation> donations;
  private Boolean superuser;
  private Boolean valid;

  public Voluntary(@Valid RequestCreateVoluntary request) {
    this.user = new User(request.username(), request.password(), request.salt());
    this.fullname = request.fullname();
    this.superuser = false;
    this.valid = true;
  }

  public void updateVoluntary(RequestUpdateVoluntary request) {
    if (request.username() != null || request.password() != null || request.salt() != null) {
      this.user.updateUser(request.username(), request.password(), request.salt());
    }
    if (request.fullname() != null) {
      this.fullname = request.fullname();
    }
  }

  public void updateVoluntary(Stand stand){
    this.stand = stand;
  }

  public void deleteVoluntary() {
    this.valid = false;
  }
}

File: frontend/src/data/volunteers/Voluntary.ts
import SummaryFunction from "./Function";

export default interface Voluntary {
  uuid: string;
  username: string;
  fullname: string;
  summaryFunction?: SummaryFunction;
  voluntaryRole: VoluntaryRole;
}

export interface SummaryVoluntary {
  uuid: string;
  fullname: string;
  summaryFunction?: SummaryFunction;
  voluntaryRole: VoluntaryRole;
}

export interface UpdateVoluntary {
  uuid: string;
  username?: string;
  password?: string;
  fullname?: string;
}

export interface UpdateVoluntaryFunction {
  uuid: string;
  functionUuid: string;
}

export interface UpdateVoluntaryRole {
  uuid: string;
  voluntaryRole: VoluntaryRole;
}

export enum VoluntaryRole {
  VOLUNTARY = "voluntary",
  MANAGEMENT = "management",
  ADMIN = "admin",
}
